### PR TITLE
ENH: `special.rel_entr`: Avoid overflow before computing log

### LINKED ISSUE #20710: ENH: special.rel_entr: avoid premature overflow
### Describe your issue.

Looks like there's opportunity for improvement in `scipy.special.rel_entr`. Presumably, it takes the ratio `x/y` before taking the logarithm because it's faster than taking the difference of two logs, but this can overflow when the correct result is finite. Perhaps a compromise is to check for the overflow and, in that case, take the time to compute the two logs separately.

This came up gh-20673 as a Hypothesis counterexample rather than a real use case, which is why I consider it an enhancement request.

### Reproducing Code Example

```python
import numpy as np
from scipy import special
x = np.asarray([2., 3., 4.])
y = np.asarray(np.finfo(np.float64).tiny)
special.rel_entr(x, y)  # array([1418.17913143, 2128.48509246,           inf])
x * (np.log(x) - np.log(y))  # array([1418.17913143, 2128.48509246, 2839.13085157])
```

### Error message

```shell
-
```

### SciPy/NumPy/Python version and system information

```shell
main
```

### PR DESCRIPTION
**Pull request by nickodell**
#### Reference issue
Fixes #20710

#### What does this implement/fix?
This PR improves the accuracy of `rel_entr()` in two ways:

1. If x / y overflows, underflows, or results in a subnormal number, the result is computed with the formula `x * (log(x) - log(y))`.
2. If x / y is close to 1, then a formula based on log1p is used instead. This improves accuracy for small rel_entr() values.

#### Performance regression

This is slower than the original `rel_entr()`.

Example benchmark script:

```
from scipy.special import rel_entr
import numpy as np
import timeit

x = np.random.rand(1_000_000)
y = np.random.rand(1_000_000)

def func():
    rel_entr(x, y)

print(timeit.timeit(func, number=100))
```

Prints 1.958 on main vs 3.587 on my branch.

#### Additional information

The log1p formula was contributed by @fancidev. 

I didn't understand why it worked, so here's a derivation of the formula:

```
rel_entry(x, y) = x * log(x / y)
rel_entry(x, y) = x * log((x - y + y) / y)
rel_entry(x, y) = x * log(1 + (x - y) / y)
rel_entry(x, y) = x * log1p((x - y) / y)
```

cc @mdhaber

### CODE REVIEW

**Comment by mdhaber** [comment]
Thanks, it would be great to avoid the over/underflow. I'll leave it to the `special` maintainers to review.

**Comment by nickodell** [comment]
Windows CI failure seems to be related to gh-20806.

```
 ______________ TestDifferentialEvolutionSolver.test_strategy_fn _______________
[gw0] win32 -- Python 3.12.3 C:\hostedtoolcache\windows\Python\3.12.3\x64\python.exe
Test passed but took too long to run: Duration 1.8958849000000555s > 1.0s
```

Looking into the [CircleCI failure](https://app.circleci.com/pipelines/github/scipy/scipy/27523/workflows/38d267a8-03ac-4ab6-a4b6-091b0a4b16a1/jobs/92817) now.

**Comment by nickodell** [comment]
Seems that this change to `rel_entr()` means that the doc test no longer passes. The new value differs by one ULP from the old value. Checking with mpmath, the new value seems to be closer to the true value.

Here's a possible fix. It is a bit ugly, though - I think it makes the docs a little less clear here. What do you think?

```
diff --git a/scipy/stats/_entropy.py b/scipy/stats/_entropy.py
index 1e6c8d9e1..79954e24c 100644
--- a/scipy/stats/_entropy.py
+++ b/scipy/stats/_entropy.py
@@ -123,8 +123,8 @@ def entropy(pk: np.typing.ArrayLike,
 
     >>> D = entropy(pk, qk, base=base)
     >>> D
-    0.7369655941662062
-    >>> D == np.sum(pk * np.log(pk/qk)) / np.log(base)
+    0.7369655941662061
+    >>> np.isclose(D, np.sum(pk * np.log(pk/qk)) / np.log(base))
     True
 
     The cross entropy can be calculated as the sum of the entropy and
```

**Comment by fancidev** [comment]
> Seems that this change to `rel_entr()` means that the doc test for `scipy.stats.entropy()` no longer passes. The new value differs by one ULP from the old value. Checking with mpmath, the new value seems to be closer to the true value.

Nice catch. What is the correctly rounded result? (I find that by `float(x)` where `x` is the result produced by `mpmath`.)

**Comment by fancidev** [comment]
> I did the same thing to round it to a float. Here's the unrounded result:
> 

Cool, this shows the new result is more accurate than the old one.

I don't know what the maintainers think, but I would remove those lines from the doc test:

```python
>>> D == np.sum(pk * np.log(pk/qk)) / np.log(base)
    True
```

For a few reasons. (1) Exact floating point equality should rarely be used for testing. (2) There is already unit tests for this function (I hope). (3) It doesn't appear to be best practice to test against a code that reproduces the expression.

**Comment by mdhaber** [comment]
Just to clarify, the primary purpose of doctests here is not testing but to demonstrate the API to the user. We do run the checks, but mostly to make sure the syntax is right and that the example is showing what we think it is. There may be other ways of doing this without an equality sign (e.g. print both results), but sometimes that is the most obvious way to make the point.

**Comment by fancidev** [comment]
The code looks good to me. (I also verified that the value of `tiny` is correctly copy-pasted.)

In addition, if `x` and `y` are both INFINITY, the function returns `nan` both before and after the PR. (Whether this is desirable/expected may be subject to discussion, but the PR retains the behaviour.)

**Comment by nickodell** [comment]
@mdhaber I see you re-triggered the test - do you need me to merge main into this branch so that the fail-slow test passes?

File: scipy/special/_convex_analysis.pxd
Comment (steppi) [inline scipy/special/_convex_analysis.pxd:28]: You can get this value through

```cython
from libc.float cimport DBL_MIN
```

so no need to define it within the function.

File: scipy/special/_convex_analysis.pxd
Comment (steppi) [inline scipy/special/_convex_analysis.pxd:46]: GitHub isn't letting me turn this into a suggestion, but I think it would be good to avoid most of the nested indentation. You could structure this as

```cython
    if isnan(x) or isnan(y):
        return NAN
    if x <= 0 or y <= 0:
        if x == 0 and y >= 0:
            return 0
        return INFINITY
    ratio = x / y
    if 0.5 < ratio < 2:
        # When x and y are close, this is more accurate
        return x * log1p((x - y) / y)
    if DBL_MIN < ratio < INFINITY:
        # There are no underflow/overflow issues
        return x * log(ratio)
    # x and y are so far apart that taking x / y
    # results in either an underflow, overflow,
    # or subnormal number. Do the logarithm first
    return x * (log(x) - log(y))
```

I replaced `tiny` with `DBL_MIN` following my other suggestion.

**Review by steppi — CHANGES_REQUESTED** [review]
Looks good @nickodell and I think the improved accuracy outweighs the increase in runtime. Just a few minor suggestions.

**Comment by mdhaber** [comment]
I would need to merge main because it turns out that test got much slower recently. But you can also just ignore the failure. I was just trying to make CI green for you.

File: scipy/special/_convex_analysis.pxd
Comment (nickodell) [reply scipy/special/_convex_analysis.pxd]: Thanks, thought there must be a better way to do this.

Fixed in 78b66f8.

File: scipy/special/_convex_analysis.pxd
Comment (nickodell) [reply scipy/special/_convex_analysis.pxd]: Fixed in 3727337.

File: scipy/stats/_entropy.py
Comment (steppi) [inline scipy/stats/_entropy.py:127]: How about using the actual value that you're getting for `D to avoid the "slight lie" and then using `np.allclose` with a very small `rtol` (around twice machine epsilon?).

Comment (nickodell) [reply scipy/stats/_entropy.py]: See 3da2b2d.

**Review by steppi — APPROVED** [review]
Looks good. Thanks @nickodell!

File: scipy/special/_convex_analysis.pxd
Comment (WarrenWeckesser) [inline scipy/special/_convex_analysis.pxd:33]: As long as you are adding code to get the edge cases correct, you might as well also handle x = ±∞ and y = ±∞.  In all four possibilities, the function should return nan.

Comment (steppi) [reply scipy/special/_convex_analysis.pxd:33]: My understanding is that handling out of domain cases by returning $\infty$ (or $-\infty$ for concave functions) was a deliberate choice in `_convex_analysis.pxd`, and this behavior is consistent across the functions defined in this file. See for instance the discussion in https://github.com/scipy/scipy/issues/3421. If we should change this behavior, I think it needs it's own discussion first, and if we proceed, the change should be done simultaneously for all functions in `_convex_analysis.pxd`. It shouldn't impact this PR.

One thing that may still be inconsistent is that when `x` and `y` are both $+\infty$, the output is `nan`, as @fancidev pointed out above. I think this can be addressed separately if needed.

**Comment by steppi** [comment]
@WarrenWeckesser, I've merged this but I'll create an issue today about edge case behavior for these convex analysis functions.

File: scipy/special/_convex_analysis.pxd
Comment (WarrenWeckesser) [reply scipy/special/_convex_analysis.pxd:33]: @steppi, I see, thanks.  I hadn't followed that discussion closely, so I was applying the usual rules that `inf/inf` and `inf - inf` return `nan`.

**Pull request merged**
